### DISCUSSION TITLE
save both original and forwarded request and response

### DISCUSSION THREAD

**Discussion by wiktortomczak** [mitmproxy/mitmproxy#5846] [General]
Hi,

What is a good way to save / write / log (request, response) pairs on both ends (client-facing, server-facing) of mitmproxy? That is, all 4 of:

* c->p: original request from the client
* p->c: final response to the client, possibly rewritten by the proxy
* p->s: request forwarded to the server and possibly rewritten by the proxy
* s->p: original response from the server

               c->p                         p->s
               original                     forwarded rewritten
               request                      request
             --------------------->        --------------> 
      client                         proxy                 server
             <---------------------        <--------------
               p->c                          s->p
               forwarded rewritten           original
               response                      response

As far as I can see, the `-w` flag (of `mitmdump`) only saves single (request, response) pairs, so either the original or the forwarded request / response are not saved (and possibly neither of those, only some intermediate stage, before / after other addons that modify request / response).

Thanks for help.

**Comment by ulope** [comment]
Did you ever find out a way to do this?

I'm having a very weird issue where a specific client can't successfully connect to another specific server. 
Introducing mitmproxy for debugging "fixed" the issue (i.e. made the request work), so it would really be helpful to see both sides of the request to compare.
